hedera-node/src/main/java/com/hedera/services/ledger/accounts/staking/RewardCalculator.java

```diff
@@ -31,95 +31,71 @@
 
 	private long rewardsPaid;
-	private long accountReward;
-	private long accountUpdatedStakePeriodStart;
 
 	@Inject
-	public RewardCalculator(final StakePeriodManager stakePeriodManager,
-			final StakeInfoManager stakeInfoManager) {
+	public RewardCalculator(final StakePeriodManager stakePeriodManager, final StakeInfoManager stakeInfoManager) {
 		this.stakePeriodManager = stakePeriodManager;
 		this.stakeInfoManager = stakeInfoManager;
 	}
 
-	public void updateRewardChanges(final MerkleAccount account, final Map<AccountProperty, Object> changes) {
-		computePendingRewards(account);
+	public void reset() {
+		rewardsPaid = 0;
+	}
 
-		if (accountReward > 0) {
+	public long computeAndApplyReward(final MerkleAccount account, final Map<AccountProperty, Object> changes) {
+		final var reward = computePendingRewards(account);
+		if (reward > 0) {
 			final var balance = finalBalanceGiven(account, changes);
-			changes.put(BALANCE, balance + accountReward);
+			changes.put(BALANCE, balance + reward);
 		}
+		rewardsPaid += reward;
+		return reward;
+	}
 
-		changes.put(STAKE_PERIOD_START, accountUpdatedStakePeriodStart);
-		rewardsPaid += accountReward; // used for adding balance change for 0.0.800 [stakingRewardAccount]
+	public long rewardsPaidInThisTxn() {
+		return rewardsPaid;
 	}
 
-	public long estimatePendingRewards(final MerkleAccount account, final MerkleStakingInfo stakingNode) {
+	public long estimatePendingRewards(final MerkleAccount account, final MerkleStakingInfo nodeStakingInfo) {
 		return computeRewardFromDetails(
 				account,
-				stakingNode,
-				stakePeriodManager.currentStakePeriod(),
+				nodeStakingInfo,
+				stakePeriodManager.estimatedCurrentStakePeriod(),
 				stakePeriodManager.effectivePeriod(account.getStakePeriodStart()));
 	}
 
-	private final void computePendingRewards(final MerkleAccount account) {
-		final var currentPeriod = stakePeriodManager.currentStakePeriod();
-
-		// Staking rewards only accumulate for a finite # of periods (currently 365 days), so get the effective start
-		var effectiveStart = stakePeriodManager.effectivePeriod(account.getStakePeriodStart());
-
-		// Check if effectiveStart is within the range for receiving rewards
-		if (stakePeriodManager.isRewardable(effectiveStart)) {
-			final long stakedNode = account.getStakedNodeAddressBookId();
-			final var stakedNodeAccount = stakeInfoManager.mutableStakeInfoFor(stakedNode);
-
-			this.accountReward = computeRewardFromDetails(account, stakedNodeAccount, currentPeriod, effectiveStart);
-			// After we've got our rewards till the last full period, it becomes our effective start
-			effectiveStart = currentPeriod - 1;
-		} else {
-			this.accountReward = 0;
-		}
-
-		this.accountUpdatedStakePeriodStart = effectiveStart;
+	private long computePendingRewards(final MerkleAccount account) {
+		return computeRewardFromDetails(
+				account,
+				stakeInfoManager.mutableStakeInfoFor(account.getStakedNodeAddressBookId()),
```

Comment:
nit : If it is not rewardable, we dont need to do `getForModify` on MerkleMap before hand?

Reply:
In general, we will have already done path replacement for every leaf in the `MerkleStakingInfo` map; so this should be very fast. 

Even in the rare case that we do an unnecessary path replacement, it won't be a problem since it will happen deterministically.